<a href="https://colab.research.google.com/github/S14vcGt/learning-notebooks/blob/main/EMDS3.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Descriptive and Inferential Statistics



## Descriptive statistics

It helps us to summarize data.

 ### Mean and Weighted Mean
 The mean is the average of a set of values. The operation is simple to do: sum
 the values and divide by the number of values. The mean is useful because it
 shows where the “center of gravity” exists for an observed set of values.
 The mean is calculated the same way for both populations and samples.

In [ ]:
 # Number of pets each person owns
sample = [1, 3, 2, 5, 7, 0, 2, 3]
mean = sum(sample) / len(sample)
mean

2.875

this in formula would be $\bar{x}=∑\frac{x_i}{n}$ for the sample mean, and $\mu=∑\frac{x_i}{N}$ for the population mean

The mean is likely familar to you, but here’s something less known about the
 mean: the mean is actually a weighted average called the weighted mean. The
 mean we commonly use gives equal importance to each value. But we can
 manipulate the mean and give each item a different weight.

 weighted mean = $\frac{(x1 ⋅w1)+(x2 ⋅w2)+(x3 ⋅w3)+...(xn ⋅wn)}{
 w1 +w2+w3+...+wn}$

 This can be helpful when we want some values to contribute to the mean more
 than others. A common example of this is weighting academic exams to give a
 final grade. If you have 3 exams and a final exam, and we give each of the 3
 exams 20% weight and the final exam 40% weight of the final grade

In [ ]:
# Three exams of .20 weight each and final exam of .40 weight
sample = [90, 80, 63, 87]
weights = [.20, .20, .20, .40]
weighted_mean = sum(s * w for s,w in zip(sample, weights)) / sum(weights)
weighted_mean

81.4

we actully don't have to use percentages with the weights since any value we use is going to be proportionalized,as long as we keep the same diferences between the values of the weights it will return the same value.

### Median

The median is the middle-most value in a set of ordered values. You
 sequentially order the values, and the median will be the center-most value. If
 you have an even number of values, you average the two center-most values.

 The median is the middle-most value in a set of ordered values. You
 sequentially order the values, and the median will be the center-most value. If you have an even number of values, you average the two center-most values.

 When your median is very different from mean, that means you have a skewed dataset with outliers.

In [ ]:
 # Number of pets each person owns
sample = [0, 1, 5, 7, 9, 10, 14]

def median(values):
  ordered = sorted(values)
  print(ordered)
  n = len(ordered)
  mid = int(n / 2) - 1 if n % 2 == 0 else int(n/2)
  if n % 2 == 0:
    return (ordered[mid] + ordered[mid+1]) / 2.0
  else:
      return ordered[mid]

print(median(sample))

[0, 1, 5, 7, 9, 10, 14]
7


There is a concept of quantiles in descriptive statistics. The concept of
 quantiles is essentially the same as a median, just cutting the data in other
 places besides the middle. The median is actually the 50% quantile, or the
 value where 50% of ordered values are behind it. Then there is the 25%,
 50%, 75% quantiles which are known as quartiles as they cut data in 25%
 increments.

### mode
 The mode is the most frequently occurring set of values. It primarily becomes
 useful when your data is repetitive and you want to find which values occur the
 most frequently.
 When no value occurs more than once, there is no mode. When two values occur
 with an equal amount of frequency, then the dataset is considered **bimodal**.


In [ ]:
# Number of pets each person owns
from collections import defaultdict
sample = [1, 3, 2, 5, 7, 0, 2, 3]
def mode(values):
  counts = defaultdict(lambda: 0)
  for s in values:
    counts[s] += 1
    max_count = max(counts.values())
    modes = [v for v in set(values) if counts[v] == max_count]
  return modes
mode(sample)

[2, 3]

 ### Variance and Standard Deviation

 The **variance** is a measure of
 how spread out our data is, in other words how far is every data point from the mean.

 For the population variance we have $$σ^2 = \frac{∑(xi−μ)^2}{
 N}$$

In [ ]:
# variance for population data
data = [0, 1, 5, 7, 9, 10, 14]
def variance(values):
  mean = sum(values) / len(values)
  _variance = sum((v - mean) ** 2 for v in values) / len(values)
  return _variance
variance(data)

21.387755102040813

The number returned for the variance is hard to relate back to our data, so we use the **standard deviation** to do so, putting the result more close to the actual data. This is the variance scaled into
a number expressed in terms of “number of pets” which makes it a bit more
 meaningful $$ σ =\sqrt{\frac{∑(xi −μ)^2}{N}}$$

In [ ]:
# standard deviation on population data
from math import sqrt
# Number of pets each person owns
data = [0, 1, 5, 7, 9, 10, 14]

def variance(values):
  mean = sum(values) / len(values)
  _variance = sum((v - mean) ** 2 for v in values) / len(values)
  return _variance

def std_dev(values):
  return sqrt(variance(values))
std_dev(data)

#### Sample Variance and standard deviation

When we average the squared differences, we divide by n-1 rather than the total number of items n. We do this to decrease any bias in a sample, and not underestimate the variance of the population based on our sample. By counting values short of 1 item in our divisor, we increase the variance and therefore capture greater uncertainty in our sample.

$$s^2 = \frac{∑(xi−\bar{x})^2}{n−1}$$


$$s= \sqrt\frac{∑(xi−\bar{x})^2}{n−1}$$

In [ ]:
from math import sqrt
# Number of pets each person owns
data = [0, 1, 5, 7, 9, 10, 14]
def variance(values, is_sample: bool = False):
  mean = sum(values) / len(values)
  _variance = sum((v - mean) ** 2 for v in values) / (len(values) - (1 if is_sample else 0))
  return _variance
def std_dev(values, is_sample: bool = False):
  return sqrt(variance(values, is_sample))

(variance(data,is_sample=True),std_dev(data, is_sample=True))

Notice that both variance and standard deviation has increased now. This is correct as a sample could be biased and imperfect representing the population. Therefore we increase the variance (and thus the standard deviation) to increase our estimate how spread out the values are. A larger variance/standard deviation shows less confidence with a larger range.

### The Normal Distribution

The normal distribution, also
 known as the Gaussian distribution, is a symmetrical bell-shaped distribution
 that has most mass around the mean, and its spread is defined as a standard
 deviation. The “tails” on either side become thinner as you move away from the
 mean. It is very helpful in both descriptive statistics and inferential statistics because allows us to treat data in a standar way, since any dataset large enough is going to be bell-shaped

  The normal distribution has several important properties that make it useful:
 - It’s symmetrical, both sides are identically mirrored at the mean which is the center

 - Most mass is at the center around the mean
 - It has a spread (being narrow or wide) that is specified by standard deviation.
 - The “tails” are the least likely outcomes, and approach zero infinitely but never touch zero.
 - It resembles a lot of phenomena in nature and daily life, and even
 generalizes non-normal problems because of the central limit theorem
 which we will talk about shortly.


### The Probability Density Function

The standard deviation plays an important role in the normal distribution,
 because it defines how “spread out” it is. It is actually one of the parameters
 alongside the mean. The probability density function (PDF) that creates the
 normal distribution is as follows:

 $f (x) = \frac{1}{σ}*\sqrt{2π}*e^{-\frac{1}{2}(\frac{x−μ^2}{σ})}$

In [ ]:
import numpy as np
import math as mt
# this was my aproach, literally the formula made code,
#which I should have known wouldn't work because of how is written in the book,
# turns out that the exponent for euler's number gets way too big,
# even using Julia (I like more Julia for numerical calculus)
#the number is interpreted as infinite, so makes sense the modifications the author
# did to the formula, he's simplifying the expression to get the same results without
# passing a so large float to exp()
def PDF(x,mean, std):
  return (mt.sqrt(2*mt.pi)*np.exp(-(x - mean**2)/2.0*std**2))/std


I_just_made_them_up= (31.56,34.44,0.897)

PDF(I_just_made_them_up[0],I_just_made_them_up[1],I_just_made_them_up[2])

# I noticed that my function maybe is more similar to the formula but also is
# utterly wrong haha, playing with the mean deliver values greater than 1, which
# is obviuslya wrong. I thought that while smaller the standard deviation greather
# is the likelihood, but experimenting with the function that does work prove me wrong
# It turns out that is a thing of the function's curve, while greater the std_dev
# greater the likelihood because there is more probabilitie acumulated under the curve
# really interesting, and obvius if you look at the curve, which I didn't

In [ ]:
import math as mt

# the one in the book, wich works and returns a likelihood
def normal_pdf(x: float, mean: float, std_dev: float) -> float:
  return (1.0 / (2.0 * mt.pi * std_dev ** 2) ** 0.5) * mt.exp(-((x - mean) ** 2 / (2.0 * std_dev ** 2)))

I_just_made_them_up= (35.56,34.44,0.897)

normal_pdf(I_just_made_them_up[0],I_just_made_them_up[1],I_just_made_them_up[2])

to understand how the PDF works I also understand better what is the cumulative density function (CDF) which I read about but didn't fully understand. The CDF does return a probability, that matches the quantiles of the PDF

With the CDF you can calculate a probability based on a value, which is very handy, and I think some ML algorithms must take advantage of this ( you get data, and with that data you get a prediction for your data)


In [ ]:
from scipy.stats import norm
mean = 64.43
std_dev = 2.99
norm.cdf(64.43, mean, std_dev)

Also with the inverse CDF you can do the opposite: giving a probability, find a value. This is usefull to generate random numbers following normal distribution ( maybe I'll see more of this in my college course of simulations and models)

In [ ]:
# so if I want the weight that 95 % of the goldens fall under
from scipy.stats import norm
norm.ppf(.95, loc=64.43, scale=2.99)

I could implemente these from scratch and I'm definetely  going to read the code, but I rather leave scratch implementations for the tasty algorithms in papers and similar 😉.

### Z-scores

 It is common to re-scale a normal distribution so that the mean is 0 and the standard deviation is 1, which is known as the standard normal distribution.

 This allows us to make it easy to compare the spread of one normal distribution to another normal distribution, even if they have different means and variances.

 Something that is of particular importance with the standard normal distribution is it expresses all x-values in terms of standard deviations, known as Z-scores.
 Turning an x-value into a Z-score uses a basic scaling formula
  $z = \frac{x−μ}{σ}$

In [ ]:
def z_score(x, mean, std):
  return (x - mean) / std
def z_to_x(z, mean, std):
  return (z * std) + mean

mean = 140000
std_dev = 3000
x = 150000
# Convert to Z-score and then back to X
z = z_score(x, mean, std_dev)
print(z)
back_to_x = z_to_x(z, mean, std_dev)
print(back_to_x)

# Inferential Statistics

>It is acceptable (perhaps even enlightened) to theorize there is no
 explanation at all and a finding is just coincidental and random.


### the central limit theorem

 states interesting once we take large enough samples of a population, calculate the mean
 of each, and plot them as a distribution:
 1. The mean of the sample means is equal to the population mean.
 2. If the population is normal, then the sample means will be normal.
 3. If the population is not normal, but the sample size is greater than 30,
 roughly the sample means will still form a normal distribution.
 4. The standard deviation of the sample means equals the population standard
 deviation divided by the square root of n

 $\text{sample standard deviation} = \frac{\text{population standard deviation}}{\sqrt{\text{sample size}}}$

  
  Let’s pretend I am measuring a population that is truly and uniformly random.
 Any value between 0.0 and 1.0 is equally likely, and no value has any
 preference. But something fascinating happens when we take increasingly large
 samples from this population, take the average of each, and then plot them into a
 histogram.

In [ ]:
 # Samples of the uniform distribution will average out to a normal distribution.
import random
import plotly.express as px
sample_size = 31
sample_count = 1000

# Central limit theorem, 1000 samples each with 31 random numbers between 0.0 and 1.0

x_values = [(sum([random.uniform(0.0, 1.0) for i in range(sample_size)])/ sample_size) for _ in range(sample_count)]

y_values = [1 for _ in range(sample_count)]

px.histogram(x=x_values, y = y_values, nbins=20).show()


 The individual numbers in the samples alone will not
 create a normal distribution. The distribution will be flat where any number is
 equally likely (known as a uniform distribution). But when we group them as
 samples and average them they form a normal distribution, and this is because of the Central Limit Theorem
#### how much sample is enough?
 While 31 is the textbook number of items you need in a sample to satisfy the
 central limit theorem and see a normal distribution, this sometimes is not the
 case. There are cases you will need an even larger sample, such as when the
 underlying distribution is asymmetrical or multimodal (meaning it has
 several peaks rather than one at the mean).
 In summary, having larger samples are better when you are uncertain of the
 underlying probability distribution. You can read more in [this article.](https://www.dummies.com/article/academics-the-arts/math/statistics/the-central-limit-theorem-whats-large-enough-142651/)

#### Confidence Interval

 A confidence interval is a range calculation showing
 how confidently we believe a sample mean (or other parameter) falls in a range
 for the population mean. In Machine Learning, it is used mainly for testing on the performance and accuracy of a model. We will be using the normal standard distribution so for this formulas to work the sample must be equal or greater than 31, according to the Central Limit Theorem. For smaller samples is needed to use the T-distribution.

For the confidence interval, we need first our parameters in the sample and a **Level Of Confidence**, which is the probability we want to prove. Then we must calculate the **critical Z-value** which is the
symmetrical range in a standard normal distribution that gives me the LOC in the center, for that we can use the inverse CDF to calculate which points containt our area of interest taking the resting probability of our LOC and splitting it so we can have the center area. Then we must calculate the  **margin of error (E)**, which is the range around the sample mean that contains the population mean at our LOC, following this:
$E=\pm z_c\frac{s}{\sqrt n}$. Applying the margin of error to the sample mean gives us the confidence interval.

In [ ]:
from scipy.stats import norm
import math as mt

def critical_z(loc):
  norm_dist = norm(loc=0.0,scale=1.0)
  left_tail = (1 - loc) /2
  right_tail = 1 - left_tail
  return norm_dist.ppf(left_tail), norm_dist.ppf(right_tail)

def confidence_interval(loc, sample_mean, sample_std_dev, n):
  lower,upper= critical_z(loc)
  error= sample_std_dev / mt.sqrt(n)
  return sample_mean + lower * error, sample_mean + upper * error

##### example

**Base on a sample of 31 golden retrievers with a sample mean of 64.408 and
a sample standard deviation of 2.05, I am 95% confident that the
population mean lies between 63.686 and 65.1296.** How do I know this?

I want to be 95% confident that my sample mean falls in the population mean range I will calculate, so that is my LOC. The critical Z value i in this case the symmetrical range in a standard normal distribution that gives me 95% probability in the center.

Logically to get 95% of the symmetrical area in the center, we would chop off the tails which have the
 remaining 5% of area. Splitting that remaining 5% area in half would give us
 2.5% area in each tail. Therefore, the areas we want to look up the x-values for
 are .025 and .975. Using the inverse CDF we can achieve this.  

 We will then return the
 corresponding lower and upper z-values containing this area. Remember, we are
 using standard normal distribution here so they will be the same other than being
 positive/negative

In [ ]:
critical_z(.95)

(-1.959963984540054, 1.959963984540054)

Okay so we get ± 1.95996 which is our critical Z-value capturing 95% of
 probability at the center of the standard normal distribution. Next we calculate the margin of error. Recall that our sample of 31 golden retrievers has a mean of 64.408 and standard deviation of 2.05.

In [ ]:
lower,upper= critical_z(.95)
error= 2.05 / mt.sqrt(31)
lower*error, upper* error

(-0.721640842980081, 0.721640842980081)

Finally, if we apply that margin of error against the sample mean, we get the confidence interval! $\text{95% confidence interval} = 64.408 ± 0.72164$

In [ ]:
confidence_interval(.95,64.408,2.05,31)

(63.68635915701992, 65.12964084298008)

 So the way to interpret this is “based on my sample of 31 golden retriever
 weights with sample mean 64.408 and sample standard deviation of 2.05, I am
 95% confident the population mean lies between 63.686 and 65.1296”. That is
 how we describe our confidence interval.

 One interesting thing to note here too is that in our margin of error formula, the
 larger n becomes the narrower our confidence interval becomes! This makes
 sense because if we have a larger sample, we are more confident in the
 population mean falling in a smaller range, hence why it’s called a confidence
 interval.

#### P-values and Hypotesis testing

A **P-value** is the probability of something occurring by chance rather than because of a hypothesized explanation.  This
 helps us frame our **null hypothesis (H0)**, saying that the variable in question
 had no impact on the experiment and any positive results is just random luck.
 The **alternative hypothesis (H1)** poses that a variable in question (called the
 controlled variable) is causing a positive result.
 Traditionally the threshold for statistical significance is a p-value of **5% or less, or .05.**. With hypothesis testing, what we are trying to achieve is to reject a null hypothesis statistically, so we can prove that a variable had an effect on our data.


#### One-tailed Test

it mesures the statistical significance only in one tail of the curve, and is useful to know in which direcction the difference go. When we approach the one-tailed test, we typically frame our null and
 alternative hypotheses using inequalities. as example, in a dataset where the mean is 18, it would be like
$$H_0 : \text{population mean} ≥ 18\\
 H_1 : \text{population mean} < 18$$
 we must calculate our p-valua for the data, and if its lower than 0.5, the null hypotesis is considered to be rejected

In [ ]:
# in this example we have a new drug given to 40 people to reduce the number
# of days that takes them to recover from flu, which in average, determined for
# another study, is 18, with a standard deviation of 1.5 and follows the norm.
#This study with the new drug gave 16 days as sample median
from scipy.stats import norm
mean = 18
std_dev = 1.5
# What x-value has 5% of area behind it?
goal_value = norm.ppf(.05, mean, std_dev)
print(goal_value)
#since the median of the study is higher than out p-values threshold
# we can't reject our null hypotesis

our_p_value = norm.cdf(16, mean, std_dev)
print(our_p_value)
# as proven, our p-value is beyond the standard

#### Two-tailed Test

 To do a two-tailed test, we frame our null and alternative hypothesis in an “equal” and “not equal” structure. This looks for determine if our variable did make any difference or not. $$H_0 : \text{population mean} = 18\\ H_1 : \text{population mean} \neq 18$$

Since our threshhold is of 5% or .05, we must split that percentage between the 2 tails, which lead us to seek significant threshhold values in 2.5% and 97.5%


In [ ]:
# keeping the example

from scipy.stats import norm

mean = 18
std_dev = 1.5
# What x-value has 2.5% of area behind it?
x1 = norm.ppf(.025, mean, std_dev)
# What x-value has 97.5% of area behind it
x2 = norm.ppf(.975, mean, std_dev)

print(f"lower limit: {x1}, upper limit: {x2}")

#so we still fail in reject the null hypotesis
#however, to calculate the p-value we must note also
#the symmetrical equivalent area on the right tail as well.
# if 16 is 2 days under the mean, the other values is going to be
# 2 days above the mean, which is 20

p1 = norm.cdf(16, mean, std_dev)

# Probability of 20 or more days
 p2 = 1.0 -  norm.cdf(20, mean, std_dev)

# P-value of both tails
 p_value = p1 + p2

 p_value

### T Distribution

The T-distribution is like a normal distribution but has fatter tails to reflect more variance and uncertainty. It is used when the data is smaller or equal to 30. An interesting thing is that while the quantity of data increase, the T distribution becomes almost exactly equal to the normal distribution.

Can be used for confidence intervals and hypothesis testing
 when you have a sample size of 30 or less. It's conceptually the same as the
 critical z-value, but we are usign a T-distribution instead of a normal distribution
 to reflect greater uncertainty. The smaller the sample size, the larger the range
 reflecting greater uncertainty,.

In [ ]:
from scipy.stats import t
# get critical value range for 95% confidence
# with a sample size of 25
n = 25
lower = t.ppf(.025, df=n-1)
upper = t.ppf(.975, df=n-1)
print(lower, upper)